In [1]:
import requests
from bs4 import BeautifulSoup
import pandas as pd
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer

print("All libraries loaded successfully!")

All libraries loaded successfully!


In [2]:
import time
import random

# Define the three New Zealand telecommunications providers and their Trustpilot URLs
providers = {
    'Spark': 'https://www.trustpilot.com/review/www.spark.co.nz',
    'One NZ': 'https://www.trustpilot.com/review/one.nz',
    '2degrees': 'https://www.trustpilot.com/review/www.2degrees.nz'
}

# Define headers to simulate a real browser visit
headers = {
    'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
}

# Test connection with Spark
response = requests.get(providers['Spark'], headers=headers)
print(f"Status code: {response.status_code}")
print("Connection successful!" if response.status_code == 200 else "Connection failed.")


Status code: 403
Connection failed.


In [3]:
pip install cloudscraper

Note: you may need to restart the kernel to use updated packages.


In [4]:
import cloudscraper

# Create a scraper instance that bypasses anti-bot protection
scraper = cloudscraper.create_scraper()

# Test connection with Spark
response = scraper.get(providers['Spark'])
print(f"Status code: {response.status_code}")
print("Connection successful!" if response.status_code == 200 else "Connection failed.")


Status code: 403
Connection failed.


In [5]:
# Test connection with Geekzone
response = requests.get('https://www.geekzone.co.nz', headers=headers)
print(f"Status code: {response.status_code}")
print("Connection successful!" if response.status_code == 200 else "Connection failed.")


Status code: 200
Connection successful!


In [1]:
# Define Geekzone search URLs for each provider
geekzone_urls = {
    'Spark': 'https://www.geekzone.co.nz/search?q=spark+nz&type=forumposts',
    'One NZ': 'https://www.geekzone.co.nz/search?q=one+nz&type=forumposts',
    '2degrees': 'https://www.geekzone.co.nz/search?q=2degrees&type=forumposts'
}

# Test fetching Spark results
response = requests.get(geekzone_urls['Spark'], headers=headers)
soup = BeautifulSoup(response.text, 'html.parser')

# Display first 500 characters to check structure
print(response.status_code)
print(soup.get_text()[:500])

NameError: name 'requests' is not defined

In [7]:
# Test Geekzone forum directly
response = requests.get('https://www.geekzone.co.nz/forums.asp', headers=headers)
soup = BeautifulSoup(response.text, 'html.parser')

print(f"Status code: {response.status_code}")
print(soup.get_text()[:1000])

Status code: 200






















Geekzone forums



















✨ 
Quic broadband | 
Apple TV | 
Samsung | 
AliExpress | 
Wise | 
Sharesies











Home
Forums
News
Reviews
Live
Search
Contact







			Welcome Guest.
		








	    You haven't logged in yet. If you don't have an account you can register now.

	  














Forums






  General forums
  






Bargains and Deals

		You can use this forum to share current bargains and deals. Please read the sticky post for special conditions. A minimum number of posts is required to create new topics.
		


1574 topics with 19666 posts
		Last post 9-Jun-2026 15:48










Finance and wealth management

		You can use this forum to discuss finance-related topics, including banking, credit card, investments, currencies, tools and platforms. This is not a registered financial advisor forum. Advice posted in discussions should not be followed without talking to your registered financial advisor.
		


1121 topics wit


In [8]:
# Search for telecommunications-related forums on Geekzone
response = requests.get('https://www.geekzone.co.nz/forums.asp', headers=headers)
soup = BeautifulSoup(response.text, 'html.parser')

# Extract all forum links and names
forums = soup.find_all('a', href=True)
for forum in forums:
    if any(word in forum.text.lower() for word in ['spark', 'one nz', '2degree', 'mobile', 'broadband', 'telecom']):
        print(f"Text: {forum.text.strip()} | URL: {forum['href']}")

Text: Quic broadband | URL: https://account.quic.nz/refer/587125
Text: Legacy mobile platforms (Windows Mobile, Palm OS, Web OS, Symbian, Meego, Blackberry) | URL: forums.asp?forumid=6
Text: Mobile handsets | URL: forums.asp?forumid=11
Text: 2degrees (including Slingshot, Orcon, Flip, Stuff Fibre, MyRepublic, 2talk and Vocus) | URL: forums.asp?forumid=81
Text: New Zealand Broadband | URL: forums.asp?forumid=49
Text: New Zealand Mobile and Wireless | URL: forums.asp?forumid=42
Text: One NZ (including Farmside) | URL: forums.asp?forumid=40
Text: Spark New Zealand (including Skinny and BigPipe) | URL: forums.asp?forumid=39
Text: Laptops and mobile computers | URL: forums.asp?forumid=37
Text: Spark Launches Verified Call to Help Combat Rising Impersonation Scams | URL: content.asp?contentid=27353


In [9]:
# Define Geekzone forum URLs for each provider
geekzone_forums = {
    'Spark': 'https://www.geekzone.co.nz/forums.asp?forumid=39',
    'One NZ': 'https://www.geekzone.co.nz/forums.asp?forumid=40',
    '2degrees': 'https://www.geekzone.co.nz/forums.asp?forumid=81'
}

# Test fetching Spark forum
response = requests.get(geekzone_forums['Spark'], headers=headers)
soup = BeautifulSoup(response.text, 'html.parser')

print(f"Status code: {response.status_code}")
print(soup.get_text()[:1000])

Status code: 200






















Forum: Spark New Zealand (including Skinny and BigPipe)




















✨ 
Quic broadband | 
Apple TV | 
Samsung | 
AliExpress | 
Wise | 
Sharesies











Home
Forums
News
Reviews
Live
Search
Contact







			Welcome Guest.
		








	    You haven't logged in yet. If you don't have an account you can register now.

	  














Forums
      
      › Spark New Zealand (including Skinny and BigPipe)














PSA Spark / Skinny / Bigpipe Fibre Broadband now supports DHCP by cbrpilot
 ...  2 3 4 5


72 replies
	Last by Steve113 on 28-Sep-2025 14:37






PSA for users of Asus Routers on the Spark Broadband network by cbrpilot


5 replies
	Last by michaelmurfy on 28-May-2024 23:38






Heads up on using 5G phones with existing SIM card by freitasm


0 replies
	Last by freitasm on 26-Feb-2021 15:48






Spark is not doing any shaping on ADSL, VDSL or Fibre. by Talkiet


1 replies
	Last by Talkiet on 28-Apr-2019 15:21






Spa


In [10]:
import time
import random

def scrape_geekzone_forum(provider_name, forum_url, max_pages=10):
    """
    Scrapes forum posts from a Geekzone provider forum.
    Returns a list of dictionaries containing post data.
    """
    all_posts = []
    
    for page in range(1, max_pages + 1):
        # Construct page URL
        page_url = f"{forum_url}&page={page}"
        response = requests.get(page_url, headers=headers)
        
        if response.status_code != 200:
            print(f"Stopped at page {page} - status code {response.status_code}")
            break
            
        soup = BeautifulSoup(response.text, 'html.parser')
        
        # Extract topic links
        topics = soup.find_all('a', href=True)
        topic_links = [t for t in topics if 'ft.asp?ftid=' in t['href']]
        
        if not topic_links:
            print(f"No more topics found at page {page}")
            break
        
        print(f"Provider: {provider_name} | Page {page} | Topics found: {len(topic_links)}")
        
        # Visit each topic and extract posts
        for topic in topic_links[:5]:  # Limit to 5 topics per page for now
            topic_url = 'https://www.geekzone.co.nz/' + topic['href']
            topic_response = requests.get(topic_url, headers=headers)
            topic_soup = BeautifulSoup(topic_response.text, 'html.parser')
            
            # Extract post text
            post_texts = topic_soup.find_all('div', class_='post-content')
            
            for post in post_texts:
                text = post.get_text(strip=True)
                if len(text.split()) >= 20:  # Minimum 20 words
                    all_posts.append({
                        'provider': provider_name,
                        'source': 'Geekzone',
                        'text': text,
                        'url': topic_url
                    })
            
            # Polite delay to avoid overloading the server
            time.sleep(random.uniform(1, 2))
        
        # Polite delay between pages
        time.sleep(random.uniform(1, 2))
    
    return all_posts

# Test with Spark forum - just 2 pages first
print("Starting test collection for Spark...")
spark_posts = scrape_geekzone_forum('Spark', geekzone_forums['Spark'], max_pages=2)
print(f"\nTotal posts collected: {len(spark_posts)}")
if spark_posts:
    print(f"\nSample post:\n{spark_posts[0]['text'][:200]}")


Starting test collection for Spark...
No more topics found at page 1

Total posts collected: 0


In [11]:
# Inspect the actual HTML structure of the Spark forum
response = requests.get(geekzone_forums['Spark'], headers=headers)
soup = BeautifulSoup(response.text, 'html.parser')

# Find all links and print those that look like topic links
all_links = soup.find_all('a', href=True)
for link in all_links[:30]:
    print(f"Text: {link.text.strip()[:50]} | href: {link['href']}")


Text: Quic broadband | href: https://account.quic.nz/refer/587125
Text: Apple TV | href: https://fave.co/4auD9HB
Text: Samsung | href: https://fave.co/3ZpXad6
Text: AliExpress | href: https://fave.co/4dsgU3n
Text: Wise | href: https://wise.com/invite/dic/mauriciof198
Text: Sharesies | href: https://sharesies.nz/r/DM4B96
Text:  | href: default.asp
Text: Home | href: default.asp
Text: Forums | href: forums.asp
Text: News | href: content.asp?contenttypeid=9
Text: Reviews | href: content.asp?contenttypeid=8
Text: Live | href: https://live.geekzone.co.nz
Text: Search | href: search.asp
Text: Contact | href: contact.asp
Text:  | href: https://www.geekzone.co.nz/register.asp
Text: logged in | href: https://www.geekzone.co.nz/login.asp
Text: register now | href: https://www.geekzone.co.nz/register.asp
Text: Forums | href: forums.asp
Text: Spark New Zealand (including Skinny and BigPipe) | href: forums.asp?forumid=39
Text:  | href: https://gadget.geekzone.co.nz/geekzone_forums_rss.asp?forumid=3

In [12]:
def scrape_geekzone_forum(provider_name, forum_url, max_pages=10):
    """
    Scrapes forum posts from a Geekzone provider forum.
    Returns a list of dictionaries containing post data.
    """
    all_posts = []
    base_url = 'https://www.geekzone.co.nz/'
    
    for page in range(1, max_pages + 1):
        page_url = f"{forum_url}&page_no={page}"
        response = requests.get(page_url, headers=headers)
        
        if response.status_code != 200:
            print(f"Stopped at page {page} - status code {response.status_code}")
            break
            
        soup = BeautifulSoup(response.text, 'html.parser')
        
        # Find topic links using correct URL pattern
        topic_links = soup.find_all('a', href=lambda x: x and 'topicid=' in x and 'page_no' not in x)
        topic_links = [t for t in topic_links if 'forumid' in t['href']]
        
        if not topic_links:
            print(f"No more topics found at page {page}")
            break
            
        print(f"Provider: {provider_name} | Page {page} | Topics found: {len(topic_links)}")
        
        for topic in topic_links[:5]:
            topic_url = base_url + topic['href']
            topic_response = requests.get(topic_url, headers=headers)
            topic_soup = BeautifulSoup(topic_response.text, 'html.parser')
            
            # Extract all text paragraphs from posts
            post_texts = topic_soup.find_all('p')
            
            for post in post_texts:
                text = post.get_text(strip=True)
                if len(text.split()) >= 20:
                    all_posts.append({
                        'provider': provider_name,
                        'source': 'Geekzone',
                        'text': text,
                        'url': topic_url
                    })
            
            time.sleep(random.uniform(1, 2))
        
        time.sleep(random.uniform(1, 2))
    
    return all_posts

# Test with Spark - 2 pages first
print("Starting test collection for Spark...")
spark_posts = scrape_geekzone_forum('Spark', geekzone_forums['Spark'], max_pages=2)
print(f"\nTotal posts collected: {len(spark_posts)}")
if spark_posts:
    print(f"\nSample post:\n{spark_posts[0]['text'][:200]}")
    

Starting test collection for Spark...
Provider: Spark | Page 1 | Topics found: 70
Provider: Spark | Page 2 | Topics found: 72

Total posts collected: 127

Sample post:
Just a quick message here to let people know that Spark, Skinny and Bigpipe now supports DHCP authentication for Fibre Broadband services (i.e. NOT for ADSL/VDSL) as well as PPP.


In [13]:
# Collect data from all three providers - 10 pages each
all_data = []

for provider, url in geekzone_forums.items():
    print(f"\nCollecting posts for {provider}...")
    posts = scrape_geekzone_forum(provider, url, max_pages=10)
    all_data.extend(posts)
    print(f"Total posts collected for {provider}: {len(posts)}")
    time.sleep(random.uniform(2, 3))

# Convert to DataFrame
df = pd.DataFrame(all_data)
print(f"\nTotal posts collected: {len(df)}")
print(f"\nPosts per provider:")
print(df['provider'].value_counts())


Provider: Spark | Page 1 | Topics found: 70
Provider: Spark | Page 2 | Topics found: 72
Provider: Spark | Page 3 | Topics found: 62
Provider: Spark | Page 4 | Topics found: 74
Provider: Spark | Page 5 | Topics found: 72
Provider: Spark | Page 6 | Topics found: 84
Provider: Spark | Page 7 | Topics found: 78
Provider: Spark | Page 8 | Topics found: 78
Provider: Spark | Page 9 | Topics found: 68
Provider: Spark | Page 10 | Topics found: 70
Total posts collected for Spark: 543

Provider: One NZ | Page 1 | Topics found: 76
Provider: One NZ | Page 2 | Topics found: 72
Provider: One NZ | Page 3 | Topics found: 70
Provider: One NZ | Page 4 | Topics found: 78
Provider: One NZ | Page 5 | Topics found: 76
Provider: One NZ | Page 6 | Topics found: 74
Provider: One NZ | Page 7 | Topics found: 78
Provider: One NZ | Page 8 | Topics found: 76
Provider: One NZ | Page 9 | Topics found: 68
Provider: One NZ | Page 10 | Topics found: 82
Total posts collected for One NZ: 359

Provider: 2degrees | Page 1 | 

In [14]:
# Save raw data to CSV
df.to_csv('geekzone_raw.csv', index=False)
print(f"Data saved successfully!")
print(f"\nDataset summary:")
print(f"Total posts: {len(df)}")
print(f"Columns: {list(df.columns)}")
print(f"\nSample of first 3 rows:")
df.head(3)

Data saved successfully!

Dataset summary:
Total posts: 1264
Columns: ['provider', 'source', 'text', 'url']

Sample of first 3 rows:


,provider,source,text,url
0,Spark,Geekzone,Just a quick message here to let people know t...,https://www.geekzone.co.nz/forums.asp?forumid=...
1,Spark,Geekzone,Which protocol you chose to use it completely ...,https://www.geekzone.co.nz/forums.asp?forumid=...
2,Spark,Geekzone,For those who wish to update their router conf...,https://www.geekzone.co.nz/forums.asp?forumid=...


In [15]:
# Basic data quality check
print("=== DATA QUALITY CHECK ===\n")

# Check for duplicates
duplicates = df.duplicated(subset=['text']).sum()
print(f"Duplicate posts: {duplicates}")

# Check text length distribution
df['word_count'] = df['text'].apply(lambda x: len(x.split()))
print(f"\nWord count statistics:")
print(df['word_count'].describe().round(1))

# Check for empty or very short texts
short_posts = df[df['word_count'] < 20]
print(f"\nPosts under 20 words (to be removed): {len(short_posts)}")

# Preview word count per provider
print(f"\nAverage word count per provider:")
print(df.groupby('provider')['word_count'].mean().round(1))

=== DATA QUALITY CHECK ===

Duplicate posts: 262

Word count statistics:
count    1264.0
mean       38.8
std        26.3
min        20.0
25%        24.0
50%        31.0
75%        43.0
max       435.0
Name: word_count, dtype: float64

Posts under 20 words (to be removed): 0

Average word count per provider:
provider
2degrees    38.7
One NZ      36.1
Spark       40.6
Name: word_count, dtype: float64


In [16]:
# Remove duplicates and reset index
df_clean = df.drop_duplicates(subset=['text']).reset_index(drop=True)

# Drop word_count column (was just for checking)
df_clean = df_clean.drop(columns=['word_count'])

# Save clean dataset
df_clean.to_csv('geekzone_clean.csv', index=False)

print("=== CLEAN DATASET SUMMARY ===\n")
print(f"Original posts: {len(df)}")
print(f"Duplicates removed: {len(df) - len(df_clean)}")
print(f"Final posts: {len(df_clean)}")
print(f"\nFinal posts per provider:")
print(df_clean['provider'].value_counts())

=== CLEAN DATASET SUMMARY ===

Original posts: 1264
Duplicates removed: 262
Final posts: 1002

Final posts per provider:
provider
Spark       440
2degrees    293
One NZ      269
Name: count, dtype: int64


In [17]:
# Define service-related keywords relevant to churn risk
service_keywords = [
    # Service quality
    'slow', 'speed', 'outage', 'down', 'disconnected', 'dropout', 'unstable',
    'connection', 'signal', 'coverage', 'reliable', 'unreliable', 'intermittent',
    # Customer experience
    'customer service', 'support', 'helpdesk', 'wait', 'hold', 'complaint',
    'frustrat', 'disappoint', 'useless', 'terrible', 'awful', 'horrible',
    # Billing
    'billing', 'charge', 'invoice', 'price', 'expensive', 'overcharged', 'refund',
    'contract', 'cancel', 'fee',
    # Switching intent
    'switch', 'leaving', 'cancel', 'moving to', 'switched to', 'left spark',
    'left one', 'left 2degrees', 'better provider', 'another provider',
    # Positive experience
    'excellent', 'great service', 'happy', 'satisfied', 'recommend', 'love'
]

# Filter posts containing at least one keyword
def contains_service_keyword(text):
    text_lower = text.lower()
    return any(keyword in text_lower for keyword in service_keywords)

df_filtered = df_clean[df_clean['text'].apply(contains_service_keyword)].reset_index(drop=True)

print(f"Original posts: {len(df_clean)}")
print(f"Filtered posts: {len(df_filtered)}")
print(f"Posts removed: {len(df_clean) - len(df_filtered)}")
print(f"\nFiltered posts per provider:")
print(df_filtered['provider'].value_counts())
print(f"\nSample post:\n{df_filtered['text'][0]}")

Original posts: 1002
Filtered posts: 393
Posts removed: 609

Filtered posts per provider:
provider
Spark       182
2degrees    114
One NZ       97
Name: count, dtype: int64

Sample post:
Just a quick message here to let people know that Spark, Skinny and Bigpipe now supports DHCP authentication for Fibre Broadband services (i.e. NOT for ADSL/VDSL) as well as PPP.


In [18]:
# Check sample posts to verify quality
for i in [0, 10, 20, 30, 40]:
    print(f"--- Post {i} | Provider: {df_filtered['provider'][i]} ---")
    print(df_filtered['text'][i][:200])
    print()


--- Post 0 | Provider: Spark ---
Just a quick message here to let people know that Spark, Skinny and Bigpipe now supports DHCP authentication for Fibre Broadband services (i.e. NOT for ADSL/VDSL) as well as PPP.

--- Post 10 | Provider: Spark ---
While Spark's network will happily support any router you wish to use (so long as it supports the minimum feature set - seehttps://www.spark.co.nz/help/internet/set-up/broadband-settings-for-third-par

--- Post 20 | Provider: Spark ---
For the majority of people 500/100Mbit is more than enough and if you're only using WiFi, don't have a homelab (as in servers hosted at home) and don't have any devices that could take up or benefit f

--- Post 30 | Provider: Spark ---
I can confirm the firmware was updated on the modem today. We now just wait a few days to announce success hopefully.

--- Post 40 | Provider: Spark ---
Honestly, your phones are old and are no longer receiving updates, and dont support the current requirement for VoLTE in NZ. Sav

In [19]:
# Save filtered dataset
df_filtered.to_csv('geekzone_filtered.csv', index=False)
print("Filtered dataset saved successfully!")
print(f"\nFinal dataset: {len(df_filtered)} posts")
print(f"Ready for NLP analysis.")


Filtered dataset saved successfully!

Final dataset: 393 posts
Ready for NLP analysis.


In [20]:
# Inspect what Trustpilot actually returns
response = scraper.get("https://www.trustpilot.com/review/www.spark.co.nz")

print(f"Status code: {response.status_code}")
print("\n--- First 3000 characters of response ---")
print(response.text[:3000])

Status code: 403

--- First 3000 characters of response ---
<!doctype html><html lang="en"><head><meta charset="UTF-8" /><meta name="viewport" content="width=device-width, initial-scale=1.0" /><title>Verifying Connection</title><link rel="stylesheet" href="/interstitial/interstitial.css" /><script src="https://a7d575be72e8.edge.sdk.awswaf.com/a7d575be72e8/b5180ee838be/challenge.js" defer></script></head><body><div class="container"><div id="loading-state"><div class="interstitial-spinner" id="spinner"></div><h1>Verifying your connection...</h1><p class="status" id="status-text">Please wait while we verify your browser.</p></div><div class="error" id="error-state"><p class="error-text" id="error-message">Verification failed. Please try again.</p></div><div class="timeout-notice" id="timeout-notice">Verification is taking longer than expected.</div><div id="button-container" class="hidden"><button class="btn btn-primary" id="retry-btn">Try Again</button></div></div><script src="/intersti

In [22]:
from playwright.sync_api import sync_playwright

with sync_playwright() as p:
    browser = p.chromium.launch(headless=False)

    page = browser.new_page()

    page.goto(
        "https://www.trustpilot.com/review/www.spark.co.nz",
        wait_until="networkidle"
    )

    page.wait_for_timeout(5000)

    print(page.content()[:5000])

    browser.close()

ModuleNotFoundError: No module named 'playwright'

In [23]:
!pip install playwright

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.3/42.3 MB 39.8 MB/s  0:00:01m0:00:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2/2 [playwright]2 [playwright]


In [24]:
import sys

!{sys.executable} -m pip install playwright
!{sys.executable} -m playwright install chromium

169.2 MiB [                    ] 0% 0.0s169.2 MiB [                    ] 0% 1249.3s169.2 MiB [                    ] 0% 889.3s169.2 MiB [                    ] 0% 1007.9s169.2 MiB [                    ] 0% 842.5s169.2 MiB [                    ] 0% 959.7s169.2 MiB [                    ] 0% 848.3s169.2 MiB [                    ] 0% 770.3s169.2 MiB [                    ] 0% 706.4s169.2 MiB [                    ] 0% 669.9s169.2 MiB [                    ] 0% 636.4s169.2 MiB [                    ] 0% 605.0s169.2 MiB [                    ] 0% 574.4s169.2 MiB [                    ] 0% 545.9s169.2 MiB [                    ] 0% 521.6s169.2 MiB [                    ] 0% 499.7s169.2 MiB [                    ] 0% 462.5s169.2 MiB [                    ] 0% 436.1s169.2 MiB [                    ] 0% 404.3s169.2 MiB [                    ] 0% 378.8s169.2 MiB [                    ] 0% 356.6s169.2 MiB [                    ] 0% 339.5s169.2 MiB [                    ] 0% 323.3s169.2 MiB [                    ] 0

In [25]:
from playwright.sync_api import sync_playwright

print("Playwright OK")

Playwright OK


In [26]:
from playwright.sync_api import sync_playwright

with sync_playwright() as p:
    browser = p.chromium.launch(headless=False)

    page = browser.new_page()

    page.goto("https://www.trustpilot.com/review/www.spark.co.nz")

    print(page.title())

    input("Press Enter to close...")

    browser.close()

Error: It looks like you are using Playwright Sync API inside the asyncio loop.
Please use the Async API instead.

In [ ]:
import asyncio
from playwright.async_api import async_playwright

async def test_trustpilot():
    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=False)

        page = await browser.new_page()

        await page.goto(
            "https://www.trustpilot.com/review/www.spark.co.nz",
            wait_until="networkidle"
        )

        print(await page.title())

        await page.screenshot(path="trustpilot_test.png")

        input("Press Enter to close...")

        await browser.close()

await test_trustpilot()

Spark New Zealand Reviews | Read Customer Service Reviews of spark.co.nz


In [2]:
# === TRUSTPILOT HTML PROCESSING ===
import glob
from datetime import datetime, timezone

# Cutoff date for the 48-month timeframe (June 2022)
CUTOFF_DATE = datetime(2022, 6, 1, tzinfo=timezone.utc)

def parse_trustpilot_html(filepath, provider):
    """
    Extracts review data from a saved Trustpilot HTML page.
    Reads the embedded __NEXT_DATA__ JSON block, which contains
    all review fields in a structured format.
    Returns a list of review dictionaries.
    """
    with open(filepath, 'r', encoding='utf-8') as f:
        html = f.read()

    soup = BeautifulSoup(html, 'html.parser')
    next_data = soup.find('script', id='__NEXT_DATA__')

    if next_data is None:
        print(f"  Warning: no __NEXT_DATA__ found in {filepath}")
        return []

    data = json.loads(next_data.string)

    try:
        reviews = data['props']['pageProps']['reviews']
    except KeyError:
        print(f"  Warning: no reviews list found in {filepath}")
        return []

    extracted = []
    for r in reviews:
        text = r.get('text')
        if not text:
            continue

        published = r['dates']['publishedDate']
        published_dt = datetime.fromisoformat(published.replace('Z', '+00:00'))

        extracted.append({
            'provider': provider,
            'source': 'Trustpilot',
            'rating': r.get('rating'),
            'title': r.get('title'),
            'text': text,
            'date': published_dt.date().isoformat(),
            'language': r.get('language')
        })

    return extracted


def process_all_pages(folder_pattern, provider):
    """
    Processes all saved HTML pages matching a filename pattern.
    Example pattern: 'trustpilot_spark_*.html'
    """
    files = sorted(glob.glob(folder_pattern))
    print(f"{provider}: found {len(files)} HTML files")

    all_reviews = []
    for filepath in files:
        reviews = parse_trustpilot_html(filepath, provider)
        all_reviews.extend(reviews)

    print(f"{provider}: extracted {len(all_reviews)} reviews total")
    return all_reviews

In [3]:
# === TRUSTPILOT HTML PROCESSING (reads from rendered article elements) ===
import glob
from datetime import datetime, timezone

# Cutoff date for the 48-month timeframe (June 2022)
CUTOFF_DATE = datetime(2022, 6, 1, tzinfo=timezone.utc)

def parse_trustpilot_html(filepath, provider):
    """
    Extracts review data from a saved Trustpilot HTML page.
    Reads from the rendered <article> elements, which correctly
    reflect the page content even after JavaScript pagination.
    Returns a list of review dictionaries.
    """
    with open(filepath, 'r', encoding='utf-8') as f:
        html = f.read()

    soup = BeautifulSoup(html, 'html.parser')
    articles = soup.find_all('article')

    extracted = []
    for art in articles:
        # A real review article contains a text paragraph with this attribute
        text_p = art.find('p', attrs={'data-service-review-text-typography': True})
        if not text_p:
            continue
        text = text_p.get_text(strip=True)

        rating_div = art.find('div', attrs={'data-service-review-rating': True})
        rating = int(rating_div['data-service-review-rating']) if rating_div else None

        title_el = art.find('h2')
        title = title_el.get_text(strip=True) if title_el else None

        time_el = art.find('time')
        date_iso = time_el.get('datetime') if time_el else None
        if not date_iso:
            continue
        published_dt = datetime.fromisoformat(date_iso.replace('Z', '+00:00'))

        extracted.append({
            'provider': provider,
            'source': 'Trustpilot',
            'rating': rating,
            'title': title,
            'text': text,
            'date': published_dt.date().isoformat(),
            'published_dt': published_dt
        })

    return extracted


def process_all_pages(folder_pattern, provider):
    """Processes all saved HTML pages matching a filename pattern."""
    files = sorted(glob.glob(folder_pattern))
    print(f"{provider}: found {len(files)} HTML files")

    all_reviews = []
    for filepath in files:
        reviews = parse_trustpilot_html(filepath, provider)
        all_reviews.extend(reviews)

    print(f"{provider}: extracted {len(all_reviews)} reviews total")
    return all_reviews

In [5]:
import requests
from bs4 import BeautifulSoup

headers = {'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36'}
url = 'https://www.geekzone.co.nz/forums.asp?forumid=39&topicid=319187'
r = requests.get(url, headers=headers, timeout=15)
print('Status:', r.status_code)

soup = BeautifulSoup(r.text, 'html.parser')

# Find every element whose class starts with _f_ and looks like a date
import re
for el in soup.find_all(attrs={'class': re.compile(r'_f_')}):
    classes = el.get('class')
    txt = el.get_text(strip=True)[:40]
    print(classes, '->', txt)
    

Status: 200
['subHeading_f_username'] -> cbrpilot
['_f_userpostcount'] -> 966
['_f_topicid'] -> 319187
['_f_topicdate'] -> 31-Mar-2025 16:58
['_f_message'] -> Hello everyone,Just a quick message here
['_f_signature'] -> My views are my own, and may not necessa
['subHeading_f_username'] -> Jiriteach
['_f_userpostcount'] -> 1141
['_f_replyid'] -> #3359171
['_f_replydate'] -> 31-Mar-2025 17:00
['_f_replymessage'] -> Awesome! - Thanks. Will help remove some
['_f_replysignature'] -> -- opinions expressed by me are solely m
['subHeading_f_username'] -> yitz
['_f_userpostcount'] -> 2286
['_f_replyid'] -> #3359173
['_f_replydate'] -> 31-Mar-2025 17:08
['_f_replymessage'] -> Does setting PPP username asNoStaticstil
['subHeading_f_username'] -> CYaBro
['_f_userpostcount'] -> 4738
['_f_replyid'] -> #3359181
['_f_replydate'] -> 31-Mar-2025 17:32
['_f_replymessage'] -> Works ok on a Northpower fibre connectio
['_f_replysignature'] -> Opinions are my own and not the views of
['subHeading_f_username'

## 1. Geekzone Re-collection with Post Dates

The original Geekzone collection captured post text and topic URLs but did not
record publication dates. To apply the 36-month temporal window consistently
across both data sources, the Geekzone posts were re-collected from their
original forum topics, capturing the publication date attached to each post.

Each forum topic contains an opening post followed by replies. The opening post
stores its date in an element of class `_f_topicdate` and its text in
`_f_message`, while each reply stores its date in `_f_replydate` and its text in
`_f_replymessage`. The extraction reads each of these post units separately, so
that every post carries its own date.

The temporal cut-off is applied at the level of the individual post rather than
the topic, because a discussion thread may remain active over several years and
a single topic can therefore contain posts written inside the window alongside
posts written outside it. Filtering by individual post date preserves
consistency with the Trustpilot data, where each review carries its own date.

In [6]:
import requests
from bs4 import BeautifulSoup
from datetime import datetime

# Standard browser header so the request is not rejected by the server
HEADERS = {
    'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) '
                  'AppleWebKit/537.36 (KHTML, like Gecko) '
                  'Chrome/124.0.0.0 Safari/537.36'
}


def parse_geekzone_topic(url, provider):
    """
    Read a single Geekzone topic page and return a list of post dictionaries.

    Each topic has one opening post and zero or more replies. The opening post
    keeps its date in '_f_topicdate' and its text in '_f_message'. Each reply
    keeps its date in '_f_replydate' and its text in '_f_replymessage'. Both
    post types are read so that every returned post carries its own date.

    Returns a list of dictionaries, each with provider, source, text, date_raw
    and url. The date is returned as raw text (for example '31-Mar-2025 16:58')
    and converted to a proper date in a later step.
    """
    response = requests.get(url, headers=HEADERS, timeout=15)
    if response.status_code != 200:
        print(f"  Skipped (status {response.status_code}): {url}")
        return []

    soup = BeautifulSoup(response.text, 'html.parser')
    posts = []

    # --- Opening post -------------------------------------------------
    # The opening date and the opening message each appear once per topic.
    opening_date = soup.find(class_='_f_topicdate')
    opening_text = soup.find(class_='_f_message')

    if opening_date and opening_text:
        posts.append({
            'provider': provider,
            'source': 'Geekzone',
            'text': opening_text.get_text(strip=True),
            'date_raw': opening_date.get_text(strip=True),
            'url': url
        })

    # --- Replies ------------------------------------------------------
    # Each reply has its own date element and its own message element. They
    # appear in the same document order, so they can be paired one to one.
    reply_dates = soup.find_all(class_='_f_replydate')
    reply_texts = soup.find_all(class_='_f_replymessage')

    for date_el, text_el in zip(reply_dates, reply_texts):
        posts.append({
            'provider': provider,
            'source': 'Geekzone',
            'text': text_el.get_text(strip=True),
            'date_raw': date_el.get_text(strip=True),
            'url': url
        })

    return posts

In [7]:
# Quick test on a single topic before running the full collection
test_posts = parse_geekzone_topic(
    'https://www.geekzone.co.nz/forums.asp?forumid=39&topicid=319187',
    'Spark'
)

print(f"Extracted {len(test_posts)} posts from the test topic")
print()
# Show the first two posts to check the fields look right
for post in test_posts[:2]:
    print('Date:', post['date_raw'])
    print('Text:', post['text'][:80])
    print('---')

Extracted 16 posts from the test topic

Date: 31-Mar-2025 16:58
Text: Hello everyone,Just a quick message here to let people know that Spark, Skinny a
---
Date: 31-Mar-2025 17:00
Text: Awesome! - Thanks. Will help remove some of the load of my UDM Pro's.Will try no
---


In [8]:
import requests
from bs4 import BeautifulSoup

# Standard browser header so the request is not rejected by the server
HEADERS = {
    'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) '
                  'AppleWebKit/537.36 (KHTML, like Gecko) '
                  'Chrome/124.0.0.0 Safari/537.36'
}


def parse_geekzone_topic(url, provider):
    """
    Read a single Geekzone topic page and return a list of post dictionaries.

    The opening post keeps its date in '_f_topicdate' and its text in
    '_f_message'. Each reply keeps its date in '_f_replydate' and its text in
    '_f_replymessage'.

    For replies, the date and the text are read together from the same reply
    block, by locating each reply date and then finding the matching message
    by its reply id. This avoids any misalignment caused by injected adverts,
    which carry a reply message but no reply date.

    Returns a list of dictionaries, each with provider, source, text, date_raw
    and url. The date is returned as raw text (for example '31-Mar-2025 16:58')
    and converted to a proper date in a later step.
    """
    response = requests.get(url, headers=HEADERS, timeout=15)
    if response.status_code != 200:
        print(f"  Skipped (status {response.status_code}): {url}")
        return []

    soup = BeautifulSoup(response.text, 'html.parser')
    posts = []

    # --- Opening post -------------------------------------------------
    opening_date = soup.find(class_='_f_topicdate')
    opening_text = soup.find(class_='_f_message')

    if opening_date and opening_text:
        posts.append({
            'provider': provider,
            'source': 'Geekzone',
            'text': opening_text.get_text(strip=True),
            'date_raw': opening_date.get_text(strip=True),
            'url': url
        })

    # --- Replies ------------------------------------------------------
    # Each genuine reply has a date element. Adverts do not, so starting from
    # the date guarantees that only real replies are collected. For each reply
    # date, the matching message is the next reply message that follows it in
    # the document, read from the same surrounding block.
    for date_el in soup.find_all(class_='_f_replydate'):
        # find_next searches forward in document order from this date element,
        # so it returns the message that belongs to this same reply.
        text_el = date_el.find_next(class_='_f_replymessage')
        if text_el is None:
            continue
        posts.append({
            'provider': provider,
            'source': 'Geekzone',
            'text': text_el.get_text(strip=True),
            'date_raw': date_el.get_text(strip=True),
            'url': url
        })

    return posts

In [9]:
# Re-test with the safer version
test_posts = parse_geekzone_topic(
    'https://www.geekzone.co.nz/forums.asp?forumid=39&topicid=319187',
    'Spark'
)

print(f"Extracted {len(test_posts)} posts from the test topic")
print()
for post in test_posts[:2]:
    print('Date:', post['date_raw'])
    print('Text:', post['text'][:80])
    print('---')

Extracted 16 posts from the test topic

Date: 31-Mar-2025 16:58
Text: Hello everyone,Just a quick message here to let people know that Spark, Skinny a
---
Date: 31-Mar-2025 17:00
Text: Awesome! - Thanks. Will help remove some of the load of my UDM Pro's.Will try no
---


## 2. Running the Full Re-collection

The list of topics to revisit is taken from the original `geekzone_filtered.csv`,
which holds the 96 topics that passed the service-relevance filter in the first
collection. Reusing this list keeps the relevance criterion unchanged, so that
the only difference introduced by the re-collection is the addition of a post
date.

Each topic is requested once, with a short pause between requests to avoid
placing unnecessary load on the server. The number of posts returned per topic
is reported during the run, so that progress can be followed and any topic that
fails to load is visible.

In [10]:
import pandas as pd
import time

# Read the original filtered file to recover the list of relevant topics.
# Each topic URL also tells us which provider it belongs to, so we keep the
# provider label alongside the URL.
original = pd.read_csv('geekzone_filtered.csv')

# Keep one row per topic URL, with its provider, removing repeated URLs.
topics = original[['provider', 'url']].drop_duplicates().reset_index(drop=True)

print(f"Topics to re-collect: {len(topics)}")
print()
print("Topics per provider:")
print(topics['provider'].value_counts())

Topics to re-collect: 96

Topics per provider:
provider
Spark       36
2degrees    31
One NZ      29
Name: count, dtype: int64


## 3. Collecting All Topics

The loop below visits each of the 96 topics in turn, extracts every dated post,
and collects them into a single list. A two-second pause separates consecutive
requests. Progress is printed for each topic, and any topic that fails to load
is reported without interrupting the run. The raw post count obtained here is
reported before any date filtering is applied, so that the effect of the
36-month restriction can be measured in the following step.

In [11]:
import time

# This list will hold every post collected across all topics.
all_geekzone_posts = []

# Walk through each topic, one row at a time.
# itertuples gives us each row with its provider and url fields.
for i, row in enumerate(topics.itertuples(index=False), start=1):
    posts = parse_geekzone_topic(row.url, row.provider)
    all_geekzone_posts.extend(posts)

    # Report progress so the run can be followed.
    print(f"[{i}/{len(topics)}] {row.provider}: {len(posts)} posts  ({row.url})")

    # Pause briefly between requests to avoid loading the server.
    time.sleep(2)

print()
print(f"Total posts collected (before date filtering): {len(all_geekzone_posts)}")

[1/96] Spark: 16 posts  (https://www.geekzone.co.nz/forums.asp?forumid=39&topicid=319187)
[2/96] Spark: 6 posts  (https://www.geekzone.co.nz/forums.asp?forumid=39&topicid=314911)
[3/96] Spark: 1 posts  (https://www.geekzone.co.nz/forums.asp?forumid=39&topicid=281579)
[4/96] Spark: 14 posts  (https://www.geekzone.co.nz/forums.asp?forumid=39&topicid=324879)
[5/96] Spark: 16 posts  (https://www.geekzone.co.nz/forums.asp?forumid=39&topicid=321481)
[6/96] Spark: 16 posts  (https://www.geekzone.co.nz/forums.asp?forumid=39&topicid=323331)
[7/96] Spark: 16 posts  (https://www.geekzone.co.nz/forums.asp?forumid=39&topicid=323419)
[8/96] Spark: 16 posts  (https://www.geekzone.co.nz/forums.asp?forumid=39&topicid=319799)
[9/96] Spark: 4 posts  (https://www.geekzone.co.nz/forums.asp?forumid=39&topicid=319775)
[10/96] Spark: 3 posts  (https://www.geekzone.co.nz/forums.asp?forumid=39&topicid=319742)
[11/96] Spark: 4 posts  (https://www.geekzone.co.nz/forums.asp?forumid=39&topicid=317936)
[12/96] Spark

## 4. Converting Dates and Applying the 36-Month Window

The posts collected so far carry their dates as text, in the format used by the
forum, for example '31-Mar-2025 16:58'. These are converted into proper date
values so that they can be compared and filtered. Posts whose date cannot be
read are reported and removed, since a post without a usable date cannot be
placed inside or outside the temporal window.

The 36-month window runs from 1 June 2023 to the collection date in June 2026.
This matches the window applied to the Trustpilot data and reflects the decision
to keep the two sources comparable, given that One NZ has existed under that name
only since the rebranding of Vodafone in April 2023. Posts dated before
1 June 2023 are removed at this stage.

In [12]:
import pandas as pd

# Turn the list of post dictionaries into a table.
geekzone = pd.DataFrame(all_geekzone_posts)

print(f"Posts before date conversion: {len(geekzone)}")

# The forum writes dates like '31-Mar-2025 16:58'. The format codes below
# describe that layout: %d day, %b short month name, %Y four-digit year,
# %H hour, %M minute. errors='coerce' turns any unreadable date into a missing
# value (NaT) instead of stopping the whole conversion.
geekzone['date'] = pd.to_datetime(
    geekzone['date_raw'],
    format='%d-%b-%Y %H:%M',
    errors='coerce'
)

# Report how many dates could not be read.
unreadable = geekzone['date'].isna().sum()
print(f"Posts with an unreadable date: {unreadable}")

# Remove posts without a usable date.
geekzone = geekzone[geekzone['date'].notna()].copy()
print(f"Posts after removing unreadable dates: {len(geekzone)}")

Posts before date conversion: 875
Posts with an unreadable date: 0
Posts after removing unreadable dates: 875


In [13]:
# The 36-month window: from 1 June 2023 up to the end of June 2026.
WINDOW_START = pd.Timestamp('2023-06-01')
WINDOW_END = pd.Timestamp('2026-06-30 23:59:59')

print(f"Posts before the 36-month window: {len(geekzone)}")
print(f"Date range in the data: {geekzone['date'].min().date()} to {geekzone['date'].max().date()}")
print()

# Keep only posts dated inside the window.
geekzone = geekzone[
    (geekzone['date'] >= WINDOW_START) &
    (geekzone['date'] <= WINDOW_END)
].copy()

print(f"Posts after the 36-month window: {len(geekzone)}")
print()
print("Posts per provider after the window:")
print(geekzone['provider'].value_counts())

Posts before the 36-month window: 875
Date range in the data: 2017-04-11 to 2026-06-08

Posts after the 36-month window: 660

Posts per provider after the window:
provider
2degrees    235
Spark       234
One NZ      191
Name: count, dtype: int64


In [14]:
# Original keyword list, unchanged.
original_keywords = [
    'slow', 'speed', 'outage', 'down', 'disconnected', 'dropout', 'unstable',
    'connection', 'signal', 'coverage', 'reliable', 'unreliable', 'intermittent',
    'customer service', 'support', 'helpdesk', 'wait', 'hold', 'complaint',
    'frustrat', 'disappoint', 'useless', 'terrible', 'awful', 'horrible',
    'billing', 'charge', 'invoice', 'price', 'expensive', 'overcharged', 'refund',
    'contract', 'cancel', 'fee',
    'switch', 'leaving', 'cancel', 'moving to', 'switched to', 'left spark',
    'left one', 'left 2degrees', 'better provider', 'another provider',
    'excellent', 'great service', 'happy', 'satisfied', 'recommend', 'love'
]

# Terms added directly, based on telecom domain knowledge.
added_keywords = [
    # Product and technology
    'broadband', 'fibre', 'mobile', 'data', 'modem', 'router', 'wifi',
    'roaming', 'sim',
    # Churn and retention
    'leave', 'port', 'porting', 'retention',
    # Negative experience
    'poor', 'worst', 'rubbish',
    # Technical fault
    'fault'
]

# The expanded list that will be used, pending the candidate test.
expanded_keywords = original_keywords + added_keywords

# Candidates held back for separate testing, because they are generic.
candidate_keywords = ['issue', 'problem', 'plan', 'error']

print(f"Original list: {len(original_keywords)} terms")
print(f"Added directly: {len(added_keywords)} terms")
print(f"Expanded list: {len(expanded_keywords)} terms")
print(f"Candidates to test separately: {candidate_keywords}")

Original list: 51 terms
Added directly: 17 terms
Expanded list: 68 terms
Candidates to test separately: ['issue', 'problem', 'plan', 'error']


In [15]:
def matches_any(text, keywords):
    """Return True if the text contains at least one keyword, lower case, substring."""
    text_lower = str(text).lower()
    return any(keyword in text_lower for keyword in keywords)


# We test against the 660 posts already inside the 36-month window.
# 'geekzone' at this point holds those 660 posts.
base = geekzone.copy()
total = len(base)

# How many posts each version retains.
keep_original = base['text'].apply(lambda t: matches_any(t, original_keywords))
keep_expanded = base['text'].apply(lambda t: matches_any(t, expanded_keywords))

print(f"Base posts in the 36-month window: {total}")
print(f"Retained by ORIGINAL list:  {keep_original.sum()}")
print(f"Retained by EXPANDED list:  {keep_expanded.sum()}")
print(f"Extra posts the expansion adds: {keep_expanded.sum() - keep_original.sum()}")
print()

# Effect of each candidate term on its own, added to the expanded list.
print("Effect of each candidate term, added individually to the expanded list:")
for term in candidate_keywords:
    keep_with = base['text'].apply(lambda t: matches_any(t, expanded_keywords + [term]))
    extra = keep_with.sum() - keep_expanded.sum()
    print(f"  + '{term}': retains {keep_with.sum()}  (adds {extra} beyond expanded)")

Base posts in the 36-month window: 660
Retained by ORIGINAL list:  310
Retained by EXPANDED list:  455
Extra posts the expansion adds: 145

Effect of each candidate term, added individually to the expanded list:
  + 'issue': retains 477  (adds 22 beyond expanded)
  + 'problem': retains 461  (adds 6 beyond expanded)
  + 'plan': retains 469  (adds 14 beyond expanded)
  + 'error': retains 456  (adds 1 beyond expanded)


In [16]:
import re

# Some keywords should match as whole words, because as loose substrings they
# would catch unrelated words. For example 'port' inside 'important' or
# 'support'. Others should match as substrings on purpose, because we want the
# word stem and all its variants. For example 'charge' should catch 'charges',
# 'charged' and 'charging', and 'frustrat' should catch 'frustrated' and
# 'frustration'. The two groups are therefore handled differently.

def matches_whole_word(text, word):
    """
    Return True if 'word' appears as a whole word in the text.

    \\b marks a word boundary, so 'port' matches 'port' but not 'important'.
    The search is case-insensitive.
    """
    pattern = r'\b' + re.escape(word) + r'\b'
    return re.search(pattern, text, flags=re.IGNORECASE) is not None


def matches_substring(text, term):
    """Return True if 'term' appears anywhere in the text, lower case."""
    return term in text.lower()


def post_is_relevant(text, whole_word_terms, substring_terms):
    """Return True if any whole-word term or any substring term is present."""
    text = str(text)
    if any(matches_whole_word(text, w) for w in whole_word_terms):
        return True
    if any(matches_substring(text, s) for s in substring_terms):
        return True
    return False

In [17]:
# Substring terms: long and distinctive, or stems we want to match in all forms.
# These are safe as substrings because their letters rarely sit inside an
# unrelated word, or because we deliberately want every variant of the stem.
substring_terms = [
    # original stems we want in all forms
    'frustrat', 'billing', 'charge', 'disconnect', 'dropout',
    'overcharged', 'invoice', 'refund',
    # original long, distinctive terms
    'outage', 'unstable', 'connection', 'coverage', 'unreliable',
    'intermittent', 'customer service', 'helpdesk', 'complaint',
    'disappoint', 'useless', 'terrible', 'horrible', 'expensive',
    'contract', 'switched to', 'moving to', 'better provider',
    'another provider', 'great service', 'excellent', 'satisfied',
    'recommend',
    # added long, distinctive product terms
    'broadband', 'fibre', 'roaming', 'modem', 'router', 'wifi',
    'porting', 'retention', 'rubbish'
]

# Whole-word terms: short words that would catch unrelated words as substrings.
whole_word_terms = [
    # original short terms
    'slow', 'speed', 'down', 'signal', 'reliable', 'support', 'wait',
    'hold', 'awful', 'price', 'fee', 'cancel', 'switch', 'leaving',
    'happy', 'love',
    # added short terms that need boundaries
    'mobile', 'data', 'sim', 'leave', 'port', 'poor', 'worst', 'fault'
]

print(f"Substring terms: {len(substring_terms)}")
print(f"Whole-word terms: {len(whole_word_terms)}")

Substring terms: 40
Whole-word terms: 24


In [18]:
# Reuse the simple substring filters from before for comparison.
def matches_any(text, keywords):
    text_lower = str(text).lower()
    return any(keyword in text_lower for keyword in keywords)


base = geekzone.copy()
total = len(base)

# Three versions to compare:
# 1. Original list, substring (what the first collection used)
# 2. Expanded list, substring (all 68 terms as loose substrings)
# 3. Two-mode list (whole-word for short terms, substring for stems)
keep_original = base['text'].apply(lambda t: matches_any(t, original_keywords))
keep_expanded = base['text'].apply(lambda t: matches_any(t, expanded_keywords))
keep_twomode = base['text'].apply(
    lambda t: post_is_relevant(t, whole_word_terms, substring_terms)
)

print(f"Base posts in the 36-month window: {total}")
print(f"1. Original list (substring):   {keep_original.sum()}")
print(f"2. Expanded list (substring):   {keep_expanded.sum()}")
print(f"3. Two-mode list:               {keep_twomode.sum()}")
print()
print(f"Two-mode adds over original:    {keep_twomode.sum() - keep_original.sum()}")
print(f"Two-mode vs expanded substring: {keep_twomode.sum() - keep_expanded.sum()}")

Base posts in the 36-month window: 660
1. Original list (substring):   310
2. Expanded list (substring):   455
3. Two-mode list:               418

Two-mode adds over original:    108
Two-mode vs expanded substring: -37


In [19]:
# Posts the two-mode filter keeps that the original list did not catch.
gained = base[keep_twomode & ~keep_original].copy()
print(f"Posts gained by the two-mode filter over the original: {len(gained)}")
print("=" * 70)

# Show a sample so the content can be inspected by eye.
for i, row in enumerate(gained.head(15).itertuples(index=False), start=1):
    print(f"[{i}] {row.provider}")
    print(row.text[:200])
    print("-" * 70)

Posts gained by the two-mode filter over the original: 130
[1] Spark
thanks, that could explain why i needed to reboot my router exactly 3 times every time i rebooted, for Spark to connect lolnot sure if this is the same issue thoughno longer using the router thankfull
----------------------------------------------------------------------
[2] Spark
freitasm: Isn't that something Quic talked about during NZNOG?Not embedded, but here's the talk:https://www.youtube.com/live/-lOF75_SxgQ?si=8OqRJIwFU-rUcvpp&t=10560The reason why this feature exists i
----------------------------------------------------------------------
[3] Spark
@AndyTyou are not following. yes hyperfibre is available now but Spark dont offer residential hyperfibre at all on any network. Be it via enable, chorus, tuatahi fibre.You need to change ISP's to one 
----------------------------------------------------------------------
[4] Spark
AndyT:?? The Enable web site and various ISP say it's now, as of 2 June, available at

In [20]:
# How many of the gained posts are caught specifically by 'data' or 'mobile'
# as whole words, to measure how much these two broad terms contribute.
for term in ['data', 'mobile', 'sim', 'port', 'fault']:
    caught = gained['text'].apply(lambda t: matches_whole_word(str(t), term))
    print(f"'{term}' (whole word) catches {caught.sum()} of the {len(gained)} gained posts")
    

'data' (whole word) catches 11 of the 130 gained posts
'mobile' (whole word) catches 17 of the 130 gained posts
'sim' (whole word) catches 16 of the 130 gained posts
'port' (whole word) catches 10 of the 130 gained posts
'fault' (whole word) catches 1 of the 130 gained posts


In [21]:
# The 37 posts the substring expansion kept but the two-mode filter rejected.
# These should be substring false positives (important, database, default...).
rejected = base[keep_expanded & ~keep_twomode].copy()
print(f"Posts rejected by two-mode that substring expansion kept: {len(rejected)}")
print("=" * 70)
for i, row in enumerate(rejected.head(12).itertuples(index=False), start=1):
    print(f"[{i}] {row.provider}")
    print(row.text[:180])
    print("-" * 70)

Posts rejected by two-mode that substring expansion kept: 37
[1] Spark
@1101Why would you not post the most important bit of information phone make and model?As above just cause you see a 4G icon when not on a call does not mean much at all what does 
----------------------------------------------------------------------
[2] Spark
wellygary:Unfortunately the only real solution is to pony up and buy a new phone....And make sure you get it from Skinny/Spark (if you want to stay with them), not a parallel-impor
----------------------------------------------------------------------
[3] Spark
Localeyes:I have just found this article which provides a link to the specific June 2022 updates for the Pixel 3 and Pixel 3A.https://www.androidpolice.com/the-pixel-3-and-3-xl-las
----------------------------------------------------------------------
[4] Spark
What are people's thoughts on this? - a Motorola G 5G Plug phone, approx 3 years old, purchased through PBTech won't work with Spark VoLTE.  Th

In [22]:
# Final two-mode keyword lists.
#
# substring_terms: the term is matched anywhere in the text. Used for long
# distinctive terms that do not sit inside unrelated words, and for stems whose
# glued variants we want to catch (switch -> switched, switching; cancel ->
# cancelled; leave -> leaving; charge -> charges, charger).
#
# whole_word_terms: the term is matched only as a standalone word, using word
# boundaries. Used for short words whose letters appear inside common unrelated
# words (sim -> simple/similar; fee -> coffee/feet; fault -> default;
# port -> important/support/portal; data -> database/update).

substring_terms = [
    # stems we want in all glued forms
    'frustrat', 'billing', 'charge', 'disconnect', 'dropout',
    'overcharged', 'invoice', 'refund', 'slow', 'speed', 'signal',
    'wait', 'cancel', 'switch', 'leave', 'price', 'poor',
    # long, distinctive terms (safe as substring)
    'outage', 'unstable', 'connection', 'coverage', 'unreliable',
    'intermittent', 'customer service', 'helpdesk', 'complaint',
    'disappoint', 'useless', 'terrible', 'horrible', 'expensive',
    'contract', 'switched to', 'moving to', 'better provider',
    'another provider', 'great service', 'excellent', 'satisfied',
    'recommend', 'reliable',
    # added product terms (long, distinctive)
    'broadband', 'fibre', 'roaming', 'modem', 'router', 'wifi',
    'porting', 'retention', 'rubbish'
]

whole_word_terms = [
    # short words that would catch unrelated words as substrings
    'down', 'hold', 'awful', 'fee', 'love', 'mobile', 'data',
    'sim', 'worst', 'fault', 'happy', 'port', 'support'
]

print(f"Substring terms: {len(substring_terms)}")
print(f"Whole-word terms: {len(whole_word_terms)}")

Substring terms: 50
Whole-word terms: 13


In [23]:
# Confirm the fix: a post with 'switched' must now be retained.
test_text = "Yup - their new $65 plan is great. Wife was paying $84 for 40gb so switched over!"
print("Caught by final two-mode filter:",
      post_is_relevant(test_text, whole_word_terms, substring_terms))

# Re-run the full count with the corrected lists.
keep_twomode_final = base['text'].apply(
    lambda t: post_is_relevant(t, whole_word_terms, substring_terms)
)
print(f"Posts retained by final two-mode filter: {keep_twomode_final.sum()}")
print()
print("Per provider:")
print(base[keep_twomode_final]['provider'].value_counts())

Caught by final two-mode filter: True
Posts retained by final two-mode filter: 427

Per provider:
provider
Spark       155
2degrees    144
One NZ      128
Name: count, dtype: int64


## 6. Saving the Re-collected Dataset

The two-mode filter is applied to the 660 posts inside the 36-month window,
retaining 427 posts that carry at least one service-relevant term. This dataset
replaces the previous `geekzone_filtered.csv`. It holds the same fields as
before, with the addition of a date column, so that it aligns with the Trustpilot
data for the consolidation step. The raw post text is kept unchanged at this
stage; cleaning and pre-processing are handled in the data preparation notebook.

In [25]:
# Apply the final two-mode filter to the windowed posts.
geekzone_final = base[keep_twomode_final].copy()

# Keep the columns needed downstream, in a clear order.
# 'date_raw' is dropped because the parsed 'date' column replaces it.
geekzone_final = geekzone_final[['provider', 'source', 'text', 'date', 'url']]

# Sort by provider then date, so the saved file is easy to scan by eye.
geekzone_final = geekzone_final.sort_values(['provider', 'date']).reset_index(drop=True)

# Save, overwriting the previous file. index=False keeps the row numbers out.
geekzone_final.to_csv('geekzone_filtered.csv', index=False)

print(f"Saved geekzone_filtered.csv with {len(geekzone_final)} posts")
print()
print("Columns:", list(geekzone_final.columns))
print()
print("Date range:", geekzone_final['date'].min(), "to", geekzone_final['date'].max())
print()
print("First few rows:")
print(geekzone_final.head())

Saved geekzone_filtered.csv with 427 posts

Columns: ['provider', 'source', 'text', 'date', 'url']

Date range: 2023-06-14 00:33:00 to 2026-06-08 08:32:00

First few rows:
   provider    source                                               text  \
0  2degrees  Geekzone  Hi I look after the SIP platform that MyRepubl...   
1  2degrees  Geekzone  Hi Kevin.Most of the above didn't make a lot o...   
2  2degrees  Geekzone  The ONT phone lines will be migrated on the da...   
3  2degrees  Geekzone  Orcon broadband how likely is it that as an ex...   
4  2degrees  Geekzone  So Update, yep switching to the 12 month plan ...   

                 date                                                url  
0 2023-08-11 19:16:00  https://www.geekzone.co.nz/forums.asp?forumid=...  
1 2023-08-11 22:43:00  https://www.geekzone.co.nz/forums.asp?forumid=...  
2 2023-08-11 23:20:00  https://www.geekzone.co.nz/forums.asp?forumid=...  
3 2023-09-03 13:26:00  https://www.geekzone.co.nz/forums.asp?forumid=..